smack-integration-test/src/main/java/org/jivesoftware/smack/roster/RosterIntegrationTest.java

```diff
@@ -104,7 +122,372 @@ private void checkIfAddedAndSubscribed(Collection<Jid> addresses) {
+        } finally {
+            rosterTwo.setSubscriptionMode(Roster.getDefaultSubscriptionMode());
+            conOne.removeAsyncStanzaListener(stanzaListener);
+        }
+    }
+
+    /**
+     * Asserts that when a user receives a presence subscription approval, the server first sends the presence stanza,
+     * followed by a roster push.
+     *
+     * @throws Exception when errors occur
+     */
+    @SmackIntegrationTest(section = "3.1.6", quote =
+        "(...)  If this check is successful, then the user's server MUST: 1. Deliver the inbound subscription " +
+        "approval to all of the user's interested resources (...). This MUST occur before sending the roster push " +
+        "described in the next step. 2. Initiate a roster push to all of the user's interested resources, containing " +
+        "an updated roster item for the contact with the 'subscription' attribute set to a value of \"to\" (...) or " +
+        "\"both\" (...).")
+    public void testReceivePresenceApprovalAndRosterPush() throws Exception {
+        IntegrationTestRosterUtil.ensureBothAccountsAreNotInEachOthersRoster(conOne, conTwo);
+
+        rosterTwo.setSubscriptionMode(Roster.SubscriptionMode.accept_all);
+
+        final ResultSyncPoint<Boolean, Exception> receivedRosterPush = new ResultSyncPoint<>();
+        final AtomicBoolean hasReceivedSubscriptionApproval = new AtomicBoolean(false);
+
+        // Replace the IQ Request Handler that processes roster pushes with one that's guaranteed to be executed
+        // synchronously with the stanza handler that listens for presence updates (below). This is to guarantee that
+        // the order in which both stanzas arrive is maintained during processing.
+        final IQRequestHandler synchronousReplacement = new AbstractIqRequestHandler(RosterPacket.ELEMENT, RosterPacket.NAMESPACE, IQ.Type.set, IQRequestHandler.Mode.sync) {
+            @Override
+            public IQ handleIQRequest(IQ iqRequest) {
+                receivedRosterPush.signal(hasReceivedSubscriptionApproval.get());
+                return IQ.createResultIQ(iqRequest);
+            }
+        };
+        final IQRequestHandler originalRequestHandler = conOne.registerIQRequestHandler(synchronousReplacement);
+
+        final StanzaListener presenceUpdateListener = stanza -> {
+            if (stanza instanceof Presence) {
+                final Presence presence = (Presence) stanza;
+                if (presence.getType() == Presence.Type.subscribed) {
+                    hasReceivedSubscriptionApproval.set(true);
+                }
+            }
+        };
+
+        // Add as a synchronous listener, again to guarantee to maintain the order of the stanzas captured by this
+        // listener and the roster pushes captured by the IQ Request Handler (above).
+        conOne.addSyncStanzaListener(presenceUpdateListener, ForEveryStanza.INSTANCE);
+
+        final Presence subscribe = PresenceBuilder.buildPresence()
```

Comment:
Please use conTwo.getStanzaFactory() to build XMPP stanzas send over a particular connection.

Reply:
`conOne`'s stanza factory, I think? The factory to be used is the one belonging to the connection that will send the stanza that is generated by it, I think?